orangecontrib/prototypes/widgets/owexplpred.py

```diff
@@ -0,0 +1,367 @@
+        )
+
+
+class Task:
+
+    future = ...
+    watcher = ...
+    canceled = False
+
+    def cancel(self):
+        self.canceled = True
+        self.future.cancel()
+        concurrent.futures.wait([self.future])
+
+class ExplainPredictions(object):
+    """
+    Class used to explain individual predictions by determining the importance of attribute values. All interactions between atributes are accounted for by calculating Shapely value.
+    Parameters:
+
+    :param data: table with dataset 
+    :type: Orange data table
+    :param model: model to be used for prediction
+    :type model: Orange model
+    :param pError: p value of error
+    :type pError: float
+    :param error: desired max error of approximation algorithm
+    :type error: float
+    :param batchSize: size of batch used in prediction, bigger batch speeds up the calculations
+    :type batchSize: int
+    :param maxIter: maximum number of iterations
+    :type maxIter: int
+    :param minIter: minimum number of iterations
+    :type minIter:int
+    """
+
+    def __init__(self, data, model, pError=0.05, error=0.05, batchSize=100, maxIter=59000, minIter=100):
+
+        self.model = model
+        self.data = data
+        self.pError = pError
+        self.error = error
+        self.batchSize = batchSize
+        self.maxIter = maxIter
+        self.minIter = minIter
+        self.atr_names = DiscreteVariable(name = 'attributes', values = [var.name for var in data.domain.attributes])
+
+
+    def anytime_explain(self, instance, callback = None):
+        dataRows, noAtr = self.data.X.shape
+        classValue = self.model(instance)[0]
+
+        # placeholders : steps, mean, sum of squared differences, calcuated contribuitons
+        steps = np.zeros((1, noAtr), dtype=float)
+        mu = np.zeros((1, noAtr), dtype=float)
+        M2 = np.zeros((1, noAtr), dtype=float)
+        expl = np.zeros((1, noAtr), dtype=float)
+        var = np.ones((1, noAtr), dtype=float)
+
+        atr_indices = np.asarray(range(noAtr)).reshape((1, noAtr))
+        batchMxSize = self.batchSize * noAtr
+        zSq = abs(st.norm.ppf(self.pError/2))**2
+
+        tiled_inst = Table.from_numpy(instance.domain, np.tile(instance.X, (self.batchSize, 1)), np.full((self.batchSize, 1), instance.Y[0]))
+        inst1 = copy.deepcopy(tiled_inst)
+        inst2 = copy.deepcopy(tiled_inst)
+        iterations_reached = np.zeros((1, noAtr))
+
+        while not(all(iterations_reached[0,:] > self.maxIter)):
+            if not(any(iterations_reached[0,:] > self.maxIter)):
+                a = np.random.choice(atr_indices[0], p=(var[0,:]/(np.sum(var[0,:]))))
+            else:
+                a = np.argmin(iterations_reached[0,:])
+
+            perm = np.random.choice([True, False], batchMxSize, replace=True)
+            perm = np.reshape(perm, (self.batchSize, noAtr))
+            rand_data = self.data.X[random.sample(
+                range(dataRows), k=self.batchSize), :]
```

Comment:
`np.random.randint(dataRows, size=self.batchSize)`

And then you don't need `random` module?